build.sbt

```
@@ -151,53 +147,96 @@ lazy val otherIncompats = Seq[ProjectReference](
+lazy val any2stringaddConversion = (project in file("incompat-3.0/any2stringadd-conversion")).settings(incompat30Settings)
+lazy val byNameParamTypeInfer = (project in file ("incompat-3.0/by-name-param-type-infer")).settings(incompat30Settings)
+lazy val defaultParamVariance = (project in file("incompat-3.0/default-param-variance")).settings(incompat30Settings)
+lazy val earlyInitializer = (project in file("incompat-3.0/early-initializer")).settings(incompat30Settings)
+lazy val existentialType = (project in file ("incompat-3.0/existential-type")).settings(incompat30Settings)
+lazy val explicitCallToUnapply = (project in file("incompat-3.0/explicit-call-to-unapply")).settings(incompat30Settings)
+lazy val javaLangEnum = (project in file("incompat-3.0/java-lang-enum")).settings(incompat30Settings)
+lazy val reflectiveCall = (project in file("incompat-3.0/reflective-call")).settings(incompat30Settings)
 
 // Contextual abstraction incompatibilities
 lazy val ambiguousConversion =
   (project in file("incompat-3.0/ambiguous-conversion"))
-    .settings(incompatSettings)
+    .settings(incompat30Settings)
     .settings(scalacOptions += "-language:implicitConversions")
-lazy val typeOfImplicitDef = (project in file("incompat-3.0/type-of-implicit-def")).settings(incompatSettings)
-lazy val accessModifier = (project in file ("incompat-3.0/access-modifier")).settings(incompatSettings)
-lazy val viewBound = (project in file("incompat-3.0/view-bound")).settings(incompatSettings)
+lazy val typeOfImplicitDef = (project in file("incompat-3.0/type-of-implicit-def")).settings(incompat30Settings)
+lazy val accessModifier = (project in file ("incompat-3.0/access-modifier")).settings(incompat30Settings)
+lazy val viewBound = (project in file("incompat-3.0/view-bound")).settings(incompat30Settings)
 
 // This one is a runtime incompatibility
 // It compiles but the runtime behavior is different
 // In this case Dotty throws an `AssertionError`
 lazy val implicitView =
   (project in file("incompat-3.0/implicit-view"))
-    .settings(runtimeIncompatSettings)
+    .settings(runtimeIncompat30Settings)
     .settings(scalacOptions += "-language:implicitConversions")
 
 
-lazy val incompatSettings = 
+/*
+  Dotty migration rewrites between 3.0 and 3.1
+  `dotc -source:3.1-migration -rewrite`
+*/
+lazy val dotty31MigrationRewrites = Seq[ProjectReference](
+  alphanumericInfix,
+  contextBoundArg,
+  etaExpansion,
+  implicitLambdaParam,
+  patternBinding,
+  wildcard
+)
+
+lazy val dotty31MigrationRewriteSettings = incompat31Settings ++ Seq(dottyRewrite := true)
+
+// This one is not a migration rewrite but a 3.1 deprecation rewrite
+// `dotc -source:3.1 -deprecation -rewrite`
+lazy val alphanumericInfix =
+  project.in(file("incompat-3.1/alphanumeric-infix"))
+    .settings(dotty31MigrationRewriteSettings)
+    .settings(scalacOptions := Seq("-source:3.1", "-deprecation", "-rewrite"))
```

I think the following would be a bit better:
```suggestion
    .settings(scalacOptions += "-deprecation")
```
to reuse the configuration already done in `dotty31MigrationRewriteSettings)`

---

I agree it is not good to overwrite the scalacOptions.

But your solution does not behave the way I want. Because, in your case, we have both the `3.1-migration` and the `deprecation` rewrites enabled. And so I cannot be sure that the rewrite is a `deprecation` rewrite and not a `migration` rewrite.

---

Ah I see. In that case, either leave it like that, or use `scalacOptions -= "-source:3.1-migration"` and then `scalacOptions ++= Seq("-source:3.1", "-deprecation")`.